File: Daily/sbgo/Leetcode48-旋转图像.go
package sbgo

//slice值传递可以被修改，slice本身可以理解成一个指针

func rotate(matrix [][]int) {
	var t = 0
	var d = len(matrix[0]) - 1
	for ; t < d; t, d = t+1, d-1 {
		rotatecircle(matrix, t, d)
	}

}

func rotatecircle(matrix [][]int, t int, d int) {
	offset := d - t
	for i := 0; i < offset; i++ {
		matrix[t][t+i], matrix[d-i][t], matrix[d][d-i], matrix[t+i][d] =
			matrix[d-i][t], matrix[d][d-i], matrix[t+i][d], matrix[t][t+i]
	}
}

File: Daily/PY/Leetcode48-旋转图像.py
from typing import List

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        l,r = 0,len(matrix) - 1
        
        def rotatecircle(m,l,r):
            off = r - l
            for i in range(off):
                m[l][l + i],m[l + i][r],m[r][r - i],m[r - i][l] = m[r - i][l],m[l][l + i],m[l + i][r],m[r][r - i]
        
        while l < r:
            rotatecircle(matrix,l,r)
            l += 1
            r -= 1